### PR TITLE
fix: [PL-32164]: Verify change of usageRestrcitions in case of SM update

### PR DESCRIPTION
**Pull request by VikasMaddukuriHarness**
## Describe your changes
We are looping through all the secrets under the sm when an sm is updated to verify its usage restrictions with the SM's restrictions. We are not checking whether the usageRestriction is updated or not and looping through each secret to match them. This is an overkill because we do this once while creating the secret. If SM usage restrictions are not updated, no need to check for the secrets again. 
This is causing high time in customer's case (NAB) where there are approx 5k secrets under one SM.
We added a check to see whether the usage restrictions are updated or not and then loop the secrets if usage is updated.

Testing:
1. Updated a secretManager (AWS-sm) with change of scope and validated whether the secrets inside this SM usage is updated or not.
2. Updated a secretManager (AWS-sm) without change of scope and validated the secrets are looped.
3. verified corner cases like without passing usageRestriction or passing it as null via api call. 
4. Added UT
## Checklist
- [x] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger utAll`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/45612)

### CODE REVIEW

**Comment by raghuAtWings** [comment]
*[`400-rest/src/main/java/software/wings/service/impl/security/SecretManagerConfigServiceImpl.java` line 125 at r1](https://reviewable.io/reviews/harness/harness-core/45612#-NRZjX7GEzmZh1Ly_1VU:-NRZjX7HAvDb9TzLw5c9:b-1qv5fc) ([raw file](https://github.com/harness/harness-core/blob/fef543f9e9d1f6a64d254882398622a3c8d68cc9/400-rest/src/main/java/software/wings/service/impl/security/SecretManagerConfigServiceImpl.java#L125)):*
> ```Java
>     } else {
>       SecretManagerConfig oldConfig = wingsPersistence.get(SecretManagerConfig.class, secretManagerConfig.getUuid());
>       if (!(oldConfig.getUsageRestrictions().getAppEnvRestrictions().equals(
> ```

shouldn't we add a unit test as well?

File: 400-rest/src/main/java/software/wings/service/impl/security/SecretManagerConfigServiceImpl.java [hunk 1]
```diff
@@ -122,8 +122,12 @@ public String save(SecretManagerConfig secretManagerConfig) {
       secretsManagerRBACService.canSetPermissions(accountId, secretManagerConfig);
     } else {
       SecretManagerConfig oldConfig = wingsPersistence.get(SecretManagerConfig.class, secretManagerConfig.getUuid());
-      secretsManagerRBACService.canChangePermissions(accountId, secretManagerConfig, oldConfig);
-      secretService.updateConflictingSecretsToInheritScopes(accountId, secretManagerConfig);
+      if (!(oldConfig.getUsageRestrictions() != null && secretManagerConfig.getUsageRestrictions() != null
+              && oldConfig.getUsageRestrictions().getAppEnvRestrictions().equals(
+                  secretManagerConfig.getUsageRestrictions().getAppEnvRestrictions()))) {
+        secretsManagerRBACService.canChangePermissions(accountId, secretManagerConfig, oldConfig);
```
Comment (nishantsaini20) [inline 400-rest/src/main/java/software/wings/service/impl/security/SecretManagerConfigServiceImpl.java:128]: Should this be out of if condition?

**Review by VikasMaddukuriHarness — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45612)* status: 2 file(s) to review. 1. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`400-rest/src/main/java/software/wings/service/impl/security/SecretManagerConfigServiceImpl.java` line 128 at r2](https://reviewable.io/reviews/harness/harness-core/45612#-NRIHonNaI7kTcwiUybk-r2-128:-NRbtuXg4z3ewt_Yz1Mg:bnv9ix1) ([raw file](https://github.com/harness/harness-core/blob/4c3a5d7e351daaf94d109c144bce9505b30c3673/400-rest/src/main/java/software/wings/service/impl/security/SecretManagerConfigServiceImpl.java#L128)):*
<details><summary><i>Previously, VikasMaddukuriHarness (VikasMaddukuriHarness) wrote…</i></summary><blockquote>

This method also does things that are related to usageRestrictions. But if there is not change in usageRestriction, we can skip validating all those things. That is why I kept this inside the if condition. We will not validate that in case there is no change in usageRestriction
</blockquote></details>

Met with nishant and he pointed out if some userPermissions are modified and if that user tries to update the Sm without change in usageRestrictions but the user does not have permissions for that usageRestrctions, the current code will not validate that and bypass it. 
So moving this check out of IF condition to cover this case.

**Review by nishantsaini20 — APPROVED** [review]

**Comment by VikasMaddukuriHarness** [comment]
trigger messagecheck

**Review by tiwaritejas — APPROVED** [review]
> Reviewed 1 of 1 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45612)* status: 2 file(s) to review. 1. unresolved discussions.

**Comment by VikasMaddukuriHarness** [comment]
trigger messagecheck

**Comment by VikasMaddukuriHarness** [comment]
SmartPRChecks-MessageCheck

**Comment by VikasMaddukuriHarness** [comment]
trigger messagecheck

**Review by nishantsaini20 — APPROVED** [review]
> Reviewed 1 of 1 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45612)* status: :shipit: complete! ✓

**Pull request merged**